### PR TITLE
Add hook for assigning general purpose variables before the core does it

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.2.x
| Description?      | This PR adds a hook to set general purpose variables on the global JavaScript object, but before assignations made from the core. The code is also slightly modified so that variables already defined are not defined again by the core. It allows customizing the content of a variable without assigning it twice, which unlock performance gains for heavy computation variables.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | CI green
| UI Tests          | 
| Fixed issue or discussion?     | none
| Related PRs       | none
| Sponsor company   | PrestaShop SA

### CODE REVIEW

File: classes/controller/FrontController.php [hunk 1]
```diff
@@ -541,31 +541,40 @@ protected function assignGeneralPurposeVariables()
...
+            'actionFrontControllerSetVariablesBefore',
+            [
+                'templateVars' => &$templateVars,
+                'cart' => $cart,
+            ],
+        );
+
+        $templateVars['cart'] = $templateVars['cart'] ?? $this->cart_presenter->present($cart);
+        $templateVars['currency'] = $templateVars['currency'] ?? $this->getTemplateVarCurrency();
+        $templateVars['customer'] = $templateVars['customer'] ?? $this->getTemplateVarCustomer();
```
Comment (matthieu-rolland) [inline classes/controller/FrontController.php:559]: All these assignations are optional: if the variable has already be defined in the hook above, we don't override it.

If the hook was not used, there's no breaking change, we set the same values as before.

**Comment by ShaiMagal** [comment]
@matthieu-rolland I tried to review, but if I click "Files changed" to see changes I see a lot of fails for example "Offset 'cart' on array{} on left side of ?? does not exist." and more. It's unreadable :(

And tests are failing as well.

**Review by Hlavtox — DISMISSED** [review]
Not sure what is the purpose of this, why would anyone completely rebuild these variables? Plus, even if they are provided here, they will be the original ones somewhere else. 🤔

**Comment by matthieu-rolland** [comment]
> @matthieu-rolland I tried to review, but if I click "Files changed" to see changes I see a lot of fails for example "Offset 'cart' on array{} on left side of ?? does not exist." and more. It's unreadable :(
> 
> And tests are failing as well.

yes phpstan doesn't like that these offest are not declared first (because they are declared in the hook), I'll fix this with declaring null values

**Comment by matthieu-rolland** [comment]
> Not sure what is the purpose of this, why would anyone completely rebuild these variables? 

let's say I want to set the cart data with my own logic... if I do it in the hook after, it will first build the cart presenter (very costly) and then do my computing in the hook after

here I can compute my own cart data in the first hook, and if it's set, it will not compute the cart presenter.

So the idea is to be able to set your own data without having twice the computing.

> Plus, even if they are provided here, they will be the original ones somewhere else. 🤔

It depends:

- They are not always computed elsewhere
- In the case of presenters, they have lazy loading, so if it is needed elsewhere, it will only compute the needed data, here the whole thing is calculated for the javascript object (because we don't know in advance what the template will need)

If you want I can only do it for presenters, maybe it makes more sense.

(I'm making this hook because we have a use case were this hook is needed.)

**Comment by Hlavtox** [comment]
@matthieu-rolland Hm hm hm. Could you expand on the usecase you need to implement? Maybe there would be other universal ways. Like you said, maybe hooking into the presenters, before they present?

**Comment by matthieu-rolland** [comment]
> @matthieu-rolland Hm hm hm. Could you expand on the usecase you need to implement? Maybe there would be other universal ways. Like you said, maybe hooking into the presenters, before they present?

To make it simple my use case is that I don't want to put the cart into the global object at all, because my theme doesn't need it on most pages. So for performance reasons, it's a big gain not to load it at all (sql and load time)

So my use case is to not load some element, but it could also be to customize what you want in it...

I think the second hook should have been before and not after

**Comment by Hlavtox** [comment]
@matthieu-rolland Oh, I understand now.

Matthieu, I suggest looking at it from a different perspective. Shouldn't we optimize the cart? Currently, with like 50 products in it, it's the biggest performance bottleneck in front office.

And if you say the cart is not needed on most pages, if it's not accessed, it shouldn't eat performance, when you made it a lazy array no?

Btw, if you need this A LOT for 8.1.x, I would much more comfortable backporting your other lazyarray PR into 8.1.x, compared to merging this PR. This looks like a workaround to me.

**Comment by matthieu-rolland** [comment]
> @matthieu-rolland Oh, I understand now.
> 
> Matthieu, I suggest looking at it from a different perspective. Shouldn't we optimize the cart? Currently, with like 50 products in it, it's the biggest performance bottleneck in front office.
> 
> And if you say the cart is not needed on most pages, if it's not accessed, it shouldn't eat performance, when you made it a lazy array no?
> 
> Btw, if you need this A LOT for 8.1.x, I would much more comfortable backporting your other lazyarray PR into 8.1.x, compared to merging this PR. This looks like a workaround to me.

The cart présenter, and all "general purpose variables" are turned into a JavaScript object with all the data in it, because we can't know in advance what the theme will need. So the benefit of lazy array is canceled.

Basically I want the possibility to remove/customize whatever I want from these "general purpose variables", prior assignment, doesn't seem like a crazy idea to me :D

What is costly is especially the price calculations, and remaking this part will be a long process, and not for 8.1.x.

**Comment by matthieu-rolland** [comment]
I mean even without talking about the cart...

 the ability to define what this global JS object will contain, without having to redefine its elements twice, is a nice feature in my opinion, so that developers can calculate the least possible elements depending on what their theme really need

it should have been like this

**Comment by kpodemski** [comment]
If you need it, let's merge it (after fixes 😀)

**Comment by matthieu-rolland** [comment]
> If you need it, let's merge it (after fixes 😀)

I'm exploring another solution right now, I'll keep you guys upated if this PR is still needed !

**Review by jolelievre — DISMISSED** [review]
Thanks @matthieu-rolland 

The purpose you described in the description and in the comments is quite clear, indeed even if we optimize the cart presenter (which you started in another PR) being able to override this array is also interesting

For example in a custom page where you don't need to use the cart you can force the cart to `[]` and the cart presenter (which is quite heavy) will never be called

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `branch` should be `develop` or `8.1.x`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

**Comment by matks** [comment]
Hi, I have changed the target branch to 8.2.x because of https://github.com/PrestaShop/PrestaShop/issues/36758

**Review by nicosomb — APPROVED** [review]

**Review by boherm — APPROVED** [review]

**Comment by kpodemski** [comment]
@Hlavtox could you re-review?

**Review by jolelievre — APPROVED** [review]

**Comment by Hlavtox** [comment]
@kpodemski I won't block it, but I don't see any value in this.

Instead of fixing the slow cart you want allow not generating it. 🤷‍♂️ Remember that third party modules and the core expects these variables to be assigned in smarty.

I understand that Prestashop SA probably has a client that is giving them hard time because of slow their new shop is, but this is not a good way to fix it. :-)))

In assignGeneralPurposeVariables you are giving the theme and module developers an offering that "these are variables that are always there, you can be sure they will be there". They are used a lot in native modules also.

**Comment by matthieu-rolland** [comment]
> @kpodemski I won't block it, but I don't see any value in this.
> 
> Instead of fixing the slow cart you want allow not generating it. 🤷‍♂️ Remember that third party modules and the core expects these variables to be assigned in smarty.
> 
> I understand that Prestashop SA probably has a client that is giving them hard time because of slow their new shop is, but this is not a good way to fix it. :-)))
> 
> In assignGeneralPurposeVariables you are giving the theme and module developers an offering that "these are variables that are always there, you can be sure they will be there". They are used a lot in native modules also.

you could already erase/replace these values after they were calculated before this PR. Now you can also set your own value before they are calculated... 

it allows caching some complex global frontend object so that it can be fetched without being calculated for example.

**Pull request merged**